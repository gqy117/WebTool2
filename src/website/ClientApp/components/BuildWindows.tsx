import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
const healthImg = require('../images/health-60to79.svg') as string;

export default class Home extends React.Component<RouteComponentProps<{}>, {}> {
    public render() {
        return <div className="widget widget-build-window sinatra/sinatra" style={{backgroundColor: "rgb(3, 160, 110)"}}>
            <h1 className="title" data-bind="name">DCE Postman Tests</h1>

            <h3 data-bind="status">Successful</h3>

            <p data-showif="show-health">
                <img data-bind-src="image" src={healthImg}></img>
                <strong data-bind="health">64</strong>% <span className="small">of recent builds passed</span>
            </p>

            <p className="context-info">
                <span data-showif="duration">Ran for <span data-bind="duration | durationFormat">an hour</span>,</span>
                <span data-showif="time" data-bind="time | dateFormat">6 days ago</span>
            </p>
            <p className="more-info"><a data-bind-href="link">View on <span data-bind="server">TeamCity</span></a></p>
            <p className="updated-at" data-bind="updatedAtMessage">Last updated at 19:31</p>
        </div>;
    }
}
