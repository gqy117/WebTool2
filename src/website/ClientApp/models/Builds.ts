export interface BuildConfigration {
    buildId: string;
    refreshInverval?: number ;
}

export interface BuildViewDTO {
    view?: BuildView;
}

export interface BuildView {
    buildId?: string;
    buildStatus?: string;
    buildName?: string;
    lastUpdated?: string;
    cssStatus: string;
    imageSrc: string;
}

export interface BuildState {
    buildState?: Build[];
}

export interface Build {
    buildId?: string;
    buildStatus?: string;
    buildName?: string;
    lastUpdated?: string;
}

export const enum ActionType {
    REQUEST_BUILD = "REQUEST_BUILD",
    RECEIVE_BUILD = "RECEIVE_BUILD",
}

export const enum BuildStatus {
    SUCCESSFUL = "Successful",
    FAILED = "Failed",
}

export const enum CssStatus {
    SUCCESSFUL = "status-successful",
    WARNING = "status-warning",
}

export interface RequestBuildsAction {
    type: ActionType.REQUEST_BUILD;
}

export interface ReceiveBuildsAction {
    type: ActionType.RECEIVE_BUILD;
    build: Build;
}

export type KnownAction = RequestBuildsAction | ReceiveBuildsAction;
