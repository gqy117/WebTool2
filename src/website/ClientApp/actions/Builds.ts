import { fetch, addTask } from "domain-task/index";
import { AppThunkAction } from "../reducers";
import * as Model from "../models/Builds";

export const actionCreators = {
    requestBuilds: (buildId?: string): AppThunkAction<Model.KnownAction> => (dispatch) => {
        let fetchTask = fetch(`api/TeamCity/Build?buildId=${buildId}`)
            .then(response => response.json() as Promise<Model.Build>)
            .then(data => {
                //console.log(data);
                dispatch({ type: Model.ActionType.RECEIVE_BUILD, build: data });
            });

        addTask(fetchTask);
        dispatch({ type: Model.ActionType.REQUEST_BUILD });
    }
};