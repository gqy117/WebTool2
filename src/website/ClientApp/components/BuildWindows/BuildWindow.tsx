import * as React from "react";
import * as Actions from "../../actions/Builds";
import { Images } from "../../images";
import * as Model from "../../models/Builds";

type BuildProps =
    Model.BuildConfigration
    & Model.BuildViewDTO
    & Model.BuildState
    & typeof Actions.actionCreators
    ;

export default class BuildWindow extends React.Component<BuildProps, {}> {
    private queryLoop: number;

    public componentWillMount() {
        // console.log(`componentWillMount`);
        this.props.requestBuilds(this.props.buildId);
    }

    public componentWillReceiveProps(nextProps: BuildProps) {
        // console.log(`componentWillReceiveProps`);
        this.startPoll();
    }

    public componentWillUnmount() {
        clearTimeout(this.queryLoop);
    }

    public startPoll() {
        clearTimeout(this.queryLoop);
        this.queryLoop = setTimeout(() => this.props.requestBuilds(this.props.buildId), this.props.refreshInverval);
    }

    public render() {
        const build = this.props.view as Model.BuildView;

        return <div>
            <div className="widget widget-build-window" style={{ backgroundColor: "rgb(3, 160, 110)" }}>
                <h1 className="title" data-bind="name"> {build.buildName} </h1>

                <h3 data-bind="status"> {build.buildStatus} </h3>

                <p data-showif="show-health">
                    <img data-bind-src="image" src={Images.health60to79}></img>
                    <strong data-bind="health">64</strong>% <span className="small">of recent builds passed</span>
                </p>

                <p className="context-info">
                    <span data-showif="duration">Ran for
                        <span data-bind="duration | durationFormat">an hour</span>
                        ,
                    </span>
                    <span data-showif="time" data-bind="time | dateFormat">6 days ago</span>
                </p>
                <p className="more-info">
                    <a data-bind-href="link">View on <span data-bind="server">TeamCity</span></a>
                </p>
                <p className="updated-at" data-bind="updatedAtMessage">Last updated at {build.lastUpdated}</p>
            </div>
        </div>;
    }
}
