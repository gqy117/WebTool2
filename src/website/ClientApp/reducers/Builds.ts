import { Reducer } from "redux/index";
import * as Model from "../models/Builds";

const initialState: Model.BuildState = { buildState: [] };

export const reducer: Reducer<Model.BuildState> = (state: Model.BuildState = initialState, action: Model.KnownAction) => {
    switch (action.type) {
        case Model.ActionType.RECEIVE_BUILD:
            const buildWithoutCurrent = (state.buildState as Model.Build[]).filter((x) => x.buildId !== action.build.buildId);

            state = {
                buildState: [
                    ...buildWithoutCurrent,
                    action.build,
                ]
            };
            break;

        case Model.ActionType.REQUEST_BUILD:
        default:
            break;
    }
    // console.log(state);
    return state;
};
