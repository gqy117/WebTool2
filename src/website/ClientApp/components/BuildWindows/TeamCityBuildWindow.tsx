import * as React from "react";
import { actionCreators } from "../../actions/Builds";
import BuildWindow from "../../containers/BuildWindows/BuildWindow";
import * as Model from "../../models/Builds";

export default class TeamCityBuildWindow extends React.Component<Model.BuildConfigration, {}> {
    public static defaultProps: Partial<Model.BuildConfigration> = {
        refreshInverval: 5 * 1000
    };

    public render() {
        return <BuildWindow
            buildId={this.props.buildId}
            refreshInverval={this.props.refreshInverval}
            startPolling={actionCreators.startPolling}
            stopPolling={actionCreators.stopPolling}
        />;
    }
}
