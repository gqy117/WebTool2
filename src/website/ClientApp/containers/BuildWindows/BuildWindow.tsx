import * as React from "react";
import { connect } from "react-redux";
import * as Actions from "../../actions/Builds";
import BuildWindow from "../../components/BuildWindows/BuildWindow";
import { Images } from "../../images";
import * as Model from "../../models/Builds";
import { ApplicationState } from "../../reducers";
import { Dictionary } from "../../utilities";

export default connect(
    (state: ApplicationState, ownProps: Model.BuildConfigration) => {
        // console.log(state);
        const build: Model.Build = (state.builds.buildState as Model.Build[]).find((x) => x.buildId === ownProps.buildId) || {};
        const buildView: Model.BuildViewDTO = new BuildWindowConverter().convertBuildView(build);

        return buildView;
    },
    Actions.actionCreators
)(BuildWindow) as typeof BuildWindow;

export class BuildWindowConverter {
    public CssStatusDic = new Dictionary<Model.CssStatus>([
        { k: Model.BuildStatus.FAILED, v: Model.CssStatus.WARNING },
        { k: Model.BuildStatus.SUCCESSFUL, v: Model.CssStatus.SUCCESSFUL },
    ]);

    public ImageDic = new Dictionary<string>([
        { k: Model.BuildStatus.FAILED, v: Images.health00to19 },
        { k: Model.BuildStatus.SUCCESSFUL, v: Images.health80plus },
    ]);

    public convertBuildView(build: Model.Build): Model.BuildViewDTO {
        return {
            view: {
                buildId: build.buildId,
                buildName: build.buildName,
                buildStatus: build.buildStatus,
                cssStatus: this.convertCssStatus(build.buildStatus),
                imageSrc: this.convertImageSrc(build.buildStatus),
                lastUpdated: this.convertLastUpdated(build.lastUpdated),
            }
        } as Model.BuildViewDTO;
    }

    public convertCssStatus(buildStatus: string = Model.BuildStatus.SUCCESSFUL): string {
        return this.CssStatusDic[buildStatus];
    }

    public convertImageSrc(buildStatus: string = Model.BuildStatus.SUCCESSFUL): string {
        return this.ImageDic[buildStatus];
    }

    public convertLastUpdated(lastUpdated?: string): string {
        return lastUpdated ? new Date(Date.parse(lastUpdated)).toLocaleTimeString() : "";
    }
}
