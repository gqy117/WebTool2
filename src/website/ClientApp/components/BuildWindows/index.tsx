import * as React from "react";
import { RouteComponentProps } from "react-router-dom";
import TeamCityBuildWindow from "./TeamCityBuildWindow";

export default class BuildWindows extends React.Component<RouteComponentProps<{}>, {}> {
    public render() {
        return <div>
            <TeamCityBuildWindow buildId={`DCEMAIN`} />
            <TeamCityBuildWindow buildId={`DCEUNITTESTS`} />
            <TeamCityBuildWindow buildId={`DCEPOSTMAN`} />
            <TeamCityBuildWindow buildId={`ADVANCEDMAIN`} />
        </div>;
    }
}
