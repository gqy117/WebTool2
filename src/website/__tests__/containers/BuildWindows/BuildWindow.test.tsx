import { BuildWindowConverter } from "../../../ClientApp/containers/BuildWindows/BuildWindow";
import * as Model from "../../../ClientApp/models/Builds";
import * as Builds from "../../consts/Builds";

describe("containers/BuildWindows/BuildWindow.tsx", () => {
    it("convertBuildView", () => {
        const converter: BuildWindowConverter = new BuildWindowConverter();
        converter.Timezone = "Pacific/Auckland";

        const buildView: Model.BuildView = converter.convertBuildView(Builds.DCE_MAIN_FAILED).view as Model.BuildView;

        expect(buildView.buildId).toBe("DCEMAIN");
        expect(buildView.buildName).toBe("DCE MAIN");
        expect(buildView.buildStatus).toBe("Failed");
        expect(buildView.cssStatus).toBe("status-warning");
        expect(buildView.imageSrc).toBe("");
    });
});
