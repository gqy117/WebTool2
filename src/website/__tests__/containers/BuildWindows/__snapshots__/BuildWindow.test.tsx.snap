// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/BuildWindows/BuildWindow.tsx convertBuildView 1`] = `
Object {
  "view": Object {
    "buildId": "DCEMAIN",
    "buildName": "DCE MAIN",
    "buildStatus": "Failed",
    "cssStatus": "status-warning",
    "imageSrc": "",
    "lastUpdated": "18:38:23",
  },
}
`;
